homeassistant/components/nzbget/__init__.py

```diff
@@ -1 +1,167 @@
+    "remaining_size": ["RemainingSizeMB", "Queue Size", "MB"],
+    "uptime": ["UpTimeSec", "Uptime", "min"],
+}
+
+CONFIG_SCHEMA = vol.Schema(
+    {
+        DOMAIN: vol.Schema(
+            {
+                vol.Required(CONF_HOST): cv.string,
+                vol.Optional(CONF_PASSWORD): cv.string,
+                vol.Optional(CONF_USERNAME): cv.string,
+                vol.Optional(CONF_PORT, default=DEFAULT_PORT): cv.port,
+                vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
+                vol.Optional(
+                    CONF_SCAN_INTERVAL, default=DEFAULT_SCAN_INTERVAL
+                ): cv.time_period,
+                vol.Optional(
+                    CONF_MONITORED_VARIABLES, default=["download_rate"]
+                ): vol.All(cv.ensure_list, [vol.In(SENSOR_TYPES)]),
+                vol.Optional(CONF_SSL, default=False): cv.boolean,
+            }
+        )
+    },
+    extra=vol.ALLOW_EXTRA,
+)
+
+
+def setup(hass, config):
+    """Set up the NZBGet sensors."""
+    host = config[DOMAIN][CONF_HOST]
+    port = config[DOMAIN].get(CONF_PORT)
+    ssl = "s" if config[DOMAIN].get(CONF_SSL) else ""
+    name = config[DOMAIN][CONF_NAME]
+    username = config[DOMAIN].get(CONF_USERNAME)
+    password = config[DOMAIN].get(CONF_PASSWORD)
+    monitored_types = config[DOMAIN].get(CONF_MONITORED_VARIABLES)
+    scan_interval = config[DOMAIN].get(CONF_SCAN_INTERVAL)
+
+    url = f"http{ssl}://{host}:{port}/jsonrpc"
+
+    try:
+        nzbget_api = NZBGetAPI(api_url=url, username=username, password=password)
+        nzbget_api.status()
+        _LOGGER.debug("Successfully validated NZBGet API connection.")
+    except (
+        requests.exceptions.ConnectionError,
+        requests.exceptions.HTTPError,
+    ) as conn_err:
+        _LOGGER.error("Error setting up NZBGet API: %s", conn_err)
+        return False
+
+    nzbget_data = hass.data[DATA_NZBGET] = NZBGetData(hass, nzbget_api)
+    nzbget_data.update()
+
+    def refresh(event_time):
+        """Get the latest data from NZBGet."""
+        nzbget_data.update()
+
+    track_time_interval(hass, refresh, scan_interval)
+
+    sensorconfig = {"sensors": monitored_types, "client_name": name}
+
+    hass.helpers.discovery.load_platform("sensor", DOMAIN, sensorconfig, config)
+
+    return True
+
+
+class NZBGetData:
+    """Get the latest data and update the states."""
+
+    def __init__(self, hass, api):
+        """Initialize the NZBGet RPC API."""
+        self.hass = hass
+        self.status = None
+        self.available = True
+        self._api = api
+
+    def update(self):
+        """Get the latest data from NZBGet instance."""
+        try:
+            self.status = self._api.status()
+            dispatcher_send(self.hass, DATA_UPDATED)
+        except requests.exceptions.ConnectionError:
+            self.available = False
+            _LOGGER.error("Unable to refresh API data")
+
+
+class NZBGetAPI:
```

Comment:
The device specific interface code, I/O and data parsing, should be extracted to a standalone library published on pypi.

https://developers.home-assistant.io/docs/en/creating_component_code_review.html#4-communication-with-devices-services

Reply:
K, I'll do that. I was wondering if this was grandfathered in, but you're right this is the time to clean it up.

Reply:
@MartinHjelmare  -- I spent some time looking at this tonight. It looks like there is an XMLRPC api that is easy to use and doesn't have you making raw http requests. There would still be some minor logic to construct the URL.  See a Python2 example here: https://nzbget.net/api/append

It looks like a couple other components use this as well instead of a PyPi library (https://github.com/home-assistant/home-assistant/blob/05ecc5a1355c7af11b5d310470a6171528dc7a2b/homeassistant/components/supervisord/sensor.py#L28).  

In your opinion, would this approach work?  If not, I can create a library for pypi, however it doesn't seem to add much over using the xmlrpc client and I'm not sure it would be more broadly useful.

Reply:
Please extract it into a library. We want all device/api protocol extracted.